vars/runATH.groovy

```diff
@@ -0,0 +1,205 @@
+                dir("deps") {
+                    if (infra.isRunningOnJenkinsInfra()) {
+                        def settingsXml = "${pwd tmp: true}/settings-azure.xml"
+                        writeFile file: settingsXml, text: libraryResource('settings-azure.xml')
+                        downloadCommand = downloadCommand + " -s ${settingsXml}"
+                    }
+                    withEnv(env) {
+                        sh downloadCommand
+                    }
+                    sh "cp jenkins-war.war jenkins.war"
+                    stash includes: 'jenkins.war', name: 'jenkinsWar'
+                }
+            } else {
+                sh("curl -o jenkins.war -L ${jenkinsURl}")
+                stash includes: '*.war', name: 'jenkinsWar'
+            }
+
+        }
+    })
+
+    ensureInNode(env, env.RUN_ATH_DOCKER_NODE ?: 'docker && highmem', {
+        if (skipExecution) {
+            return
+        }
+        stage("Running ATH") {
+            dir("athSources") {
+                unstash name: "athSources"
+                def uid = sh(script: "id -u", returnStdout: true)
+                def gid = sh(script: "id -g", returnStdout: true)
+                athContainerImage = docker.build('jenkins/ath', "--build-arg=uid='$uid' --build-arg=gid='$gid' -f src/main/resources/ath-container/Dockerfile .")
+            }
+
+            def testsToRun = metadata.tests?.join(",")
+            def categoriesToRun = metadata.categories?.join(",")
+            def browsers = metadata.browsers ?: ["firefox"]
+            def failFast = metadata.failFast ?: false
+            def rerunCount = metadata.rerunFailingTestsCount ?: 0
+            def localSnapshots = metadata.useLocalSnapshots ?: true
+
+            if (testsToRun == null && categoriesToRun == null) {
+                categoriesToRun = defaultCategory
+            }
+
+            def testingbranches = ["failFast": failFast]
+            for (browser in browsers) {
+                if (supportedBrowsers.contains(browser)) {
+
+                    def currentBrowser = browser
+                    def containerArgs = "-v /var/run/docker.sock:/var/run/docker.sock -e LOCAL_SNAPSHOTS=${localSnapshots} -e SHARED_DOCKER_SERVICE=true -e EXERCISEDPLUGINREPORTER=textfile -u ath-user"
+                    def commandBase = "./run.sh ${currentBrowser} ./jenkins.war -Dmaven.test.failure.ignore=true -DforkCount=1 -B -Dsurefire.rerunFailingTestsCount=${rerunCount}"
+                    if (infra.isRunningOnJenkinsInfra()) {
+                        def settingsXml = "${pwd tmp: true}/settings-azure.xml"
+                        writeFile file: settingsXml, text: libraryResource('settings-azure.xml')
+                        commandBase = commandBase + " -s ${settingsXml}"
+                    }
+
+                    if (testsToRun) {
```

Comment:
These two sections seem very similar. And you have flow control code interspersed with the code that runs the code that does runs the scenario. It would be much more clear if you pull out the code that runs the scenario into a method that would be more clear an easier to test. However, I am less worried about this as some of the other issues.

Reply:
+1 on not replicating this logic

Reply:
Done